### PR TITLE
Cfq vmap decoder

### PR DESCRIPTION
**Pull request by Ranapop**
With these changes the seq2tree modules were prepared to make future changes simpler, but for now the code still runs in a seq2seq setting.

`SyntaxBasedDecoder` was modified and is now different from `Decoder` in two ways:
* a `for` loop is used accross the time steps instead of `scan_in_dim`
* the code in the `SyntaxBasedDecoder` is written at batch level and it's **vmapped** inside `Seq2tree`

The new modules were **tested** in the seq2seq setting (on a single gpu machine, with batch size 256, for 1500 steps and evaluatione done every 100 steps). The results for the old and new modules are:
* old modules
  * time until first result: 4 min (dataset loading & jitting)
  * time per 100 steps: ~20 s
  * accuracy: ~50% (run1 44%, run2 55%)
* new modules:
  * time until first result: 20 min
  * time per 100 steps: ~20 s
  * accuracy: 50% (only one run)

### CODE REVIEW

**Comment by Ranapop** [comment]
First off, with these changes the setup is done for the syntax-based solution:
* the `SyntaxBasedDecoder` and `Seq2tree` modules were created (initially copies of `Decoder` and `Seq2seq`)
* the `train_syntax_based.py` is created
  * mostly with the content of `train.py` for now, but using `Seq2tree` for model creation

File: examples/cfq/models.py [hunk 1]
```diff
@@ -186,7 +203,7 @@ def apply(self,
     Args
       init_states: states to initialize the decoder hidden states (coming from
         the encoder). This assumes the encoder and decoder have the same number
-        of layers [num_layers, batch_size, hidden_size]
+        of layers. [num_layers, 2, batch_size, hidden_size]
```
Comment (marcvanzee) [inline examples/cfq/models.py:206]: What is 2?

**Review by marcvanzee — CHANGES_REQUESTED** [review]
It is difficult to review this PR since you copy a lot of code that you also change. Would it easily be possible to have 1 PR that just copies the code and 1 PR that does the vmapping? Right now it is very difficult to see how you actually change the original code so that it uses vmapping.

File: examples/cfq/models.py (hunk 1 above)
Comment (Ranapop) [reply examples/cfq/models.py:206]: An LSTM cell state is a tuple (c,h)

**Comment by Ranapop** [comment]
PR will be split into multiple PRs

File: examples/cfq/models.py [hunk 2]
```diff
@@ -42,7 +42,8 @@ def apply(self,
             projected_keys: jnp.ndarray,
             values: jnp.ndarray,
             mask: jnp.ndarray,
-            hidden_size: int = None) -> jnp.ndarray:
+            hidden_size: int = None,
+            batch_level: bool = True) -> jnp.ndarray:
```
Comment (marcvanzee) [inline examples/cfq/models.py]: Maybe rename to `use_batch_axis`?

File: examples/cfq/models.py [hunk 3]
```diff
@@ -78,8 +82,13 @@ def apply(self,
     scores = jnp.where(mask, scores, -jnp.inf)  # Using exp(-inf) = 0 below.
     scores = nn.softmax(scores, axis=-1)
 
-    context = jnp.sum(jnp.expand_dims(scores, 2) * values, axis=1)
-    return context, scores  # Shape: <float32>[batch_size, seq_len]
+    if batch_level:
+      # Shape: <float32>[batch_size, seq_len]
+      context = jnp.sum(jnp.expand_dims(scores, 2) * values, axis=1)
```
Comment (marcvanzee) [inline examples/cfq/models.py:87]: Why do we have to expend 2 dimensions if we use a batch axis and 1 otherwise? What is the intuition behind this?

File: examples/cfq/models.py [hunk 4]
```diff
@@ -377,93 +386,88 @@ def create_dropout_masks(num_masks: int, shape: Tuple,
...
         num_masks=num_layers,
-        shape=(batch_size, hidden_size),
+        shape=(hidden_size),
         dropout_rate=horizontal_dropout_rate,
         train=train)
 
-    def decode_step_fn(carry, x):
-      rng, previous_states, last_prediction, prev_attention = carry
+    time_steps = inputs.shape[0]
+    initial_h = init_states[-1,1,:]
```
Comment (marcvanzee) [inline examples/cfq/models.py]:
```suggestion
    initial_h = init_states[-1, 1, :]
```

**Review by marcvanzee — APPROVED** [review]
Looks great! Very cool :-).

File: examples/cfq/models.py (hunk 3 above)
Comment (Ranapop) [reply examples/cfq/models.py:87]: The third dimension/axis is expanded instead of the second one (when the data is at batch level, the batch is at the first axis - 0, and the second axis - 1 becomes the third axis -2)

**Review by bohnetbd — APPROVED** [review]
looks reasonable to me. Thanks!
